File: web/ts/helpers/profile-sparsify.ts
/** Sparse-profile helpers (#403 Part B).
 *
 *  We persist only what differs from the default: on save the settings page drops
 *  any advisory param or engine method equal to its default. Absence resolves to
 *  the same default server-side (advisory params via
 *  `evaluate_all`'s `{**catalog_defaults, **user_params}` merge; engine methods
 *  via `ENGINE_METHOD_DEFAULTS`), so pruning is lossless for grading and lets an
 *  already-dense profile shrink. These are pure so they can be unit-tested away
 *  from the DOM. */

import type { AdvisoryCatalogEntry } from '../types/advisories';

/** Lookup of catalog default keyed `advisoryId:paramKey`. */
export function buildParamDefaults(catalog: AdvisoryCatalogEntry[]): Map<string, number> {
  const m = new Map<string, number>();
  for (const entry of catalog) {
    for (const p of entry.parameters) m.set(`${entry.id}:${p.key}`, p.default);
  }
  return m;
}

/** Drop any advisory param whose value equals the catalog default. Params with no
 *  catalog default (unknown key) are kept. Enable flags are NOT touched here —
 *  they are deliberately never sparsified (#402: `fronts` false equals the catalog
 *  default yet is a meaningful override). */
export function pruneAdvisoryParams(
  rawParams: Record<string, Record<string, number>>,
  paramDefaults: Map<string, number>,
): Record<string, Record<string, number>> {
  const out: Record<string, Record<string, number>> = {};
  for (const advId of Object.keys(rawParams)) {
    for (const key of Object.keys(rawParams[advId])) {
      const val = rawParams[advId][key];
      const dflt = paramDefaults.get(`${advId}:${key}`);
      if (dflt !== undefined && val === dflt) continue;  // equals default → prune
      if (!out[advId]) out[advId] = {};
      out[advId][key] = val;
    }
  }
  return out;
}

/** Map an engine method to `null` ("follow the default") when it equals its
 *  declared default, else return it unchanged. Sent as an explicit null so the
 *  preview endpoint (which keys on JSON presence) grades on the resolved default
 *  and the save deletes the stored key. */
export function pruneEngineMethod(value: string, dflt: string | undefined): string | null {
  return dflt !== undefined && value === dflt ? null : value;
}

/** advisoryId → { oldKey: newKey } for the #571 extent-parameter consolidation.
 *  The lockstep sibling of `weatherbrief/analysis/advisories/extent_param_migration.py`
 *  — the two MUST move together, or a client would re-save under an old key a
 *  profile the server just migrated. Kept as a literal because the catalog no
 *  longer carries the old names: this is the only record of what each key used
 *  to be called. */
export const EXTENT_KEY_RENAMES: Record<string, Record<string, string>> = {
  cloud_top: { pct_amber: 'extent_pct_amber' },
  convective: {
    affected_pct_amber: 'extent_pct_amber',
    affected_pct_red: 'extent_pct_red',
  },
  dd_nwp_agreement: { amber_pct: 'extent_pct_amber', red_pct: 'extent_pct_red' },
  enroute_precip: {
    snow_pct_amber: 'extent_pct_amber',
    snow_moderate_pct_red: 'extent_pct_red',
  },
  fiki_icing: {
    clear_cruise_amber_pct: 'extent_pct_amber',
    clear_cruise_red_pct: 'extent_pct_red',
  },
  freezing_precip: { primed_pct_amber: 'extent_pct_amber' },
  icing_escape: {
    icing_coverage_pct_amber: 'extent_pct_amber',
    no_escape_pct_red: 'extent_pct_red',
    route_pct_amber: 'extent_pct_amber',
    min_route_pct: 'extent_pct_red',
  },
  ifr_feasibility: {
    icing_pct_amber: 'extent_pct_amber',
    icing_pct_red: 'extent_pct_red',
  },
  model_agreement: {
    poor_pct_amber: 'extent_pct_amber',
    poor_pct_red: 'extent_pct_red',
  },
  turbulence: { route_pct_amber: 'extent_pct_amber' },
  vfr_feasibility: {
    imc_pct_amber: 'extent_pct_amber',
    imc_pct_red: 'extent_pct_red',
  },
  vmc_cruise: { bkn_pct_amber: 'extent_pct_amber', ovc_pct_red: 'extent_pct_red' },
};

/** Keys whose stored VALUE flips with the name. `fiki_icing` expressed its
 *  thresholds as a percentage of the *clear* cruise, compared with `<` — the one
 *  gate that read the other way. "Amber below 70% clear" means "amber at or above
 *  30% affected", so the number must invert or the pilot's tuning would too. */
const INVERTED_PCT_KEYS: Record<string, Set<string>> = {
  fiki_icing: new Set(['clear_cruise_amber_pct', 'clear_cruise_red_pct']),
};

/** Aliases applied only when the primary name is absent, mirroring the read-path
 *  fallback chain `icing_escape` used to have (`min_route_pct` was never even a
 *  catalog key). */
const SECONDARY: Record<string, Set<string>> = {
  icing_escape: new Set(['route_pct_amber', 'min_route_pct']),
};

/** Rewrite pre-consolidation extent keys to the consolidated ones (#571 Stage 3).
 *
 *  Run this on load, before rendering: a profile the server has not migrated yet
 *  (or one restored from a stale client cache) would otherwise show the pilot
 *  their tuning missing while the old key sat in the payload doing nothing.
 *  `pruneAdvisoryParams` cannot help — it deliberately KEEPS any key it cannot
 *  prove is a default, so an unknown key lingers forever.
 *
 *  Pure: returns a new object, never mutates. A value already under the new key
 *  wins and the old key is dropped. */
export function renameExtentParams(
  rawParams: Record<string, Record<string, number>>,
): Record<string, Record<string, number>> {
  const out: Record<string, Record<string, number>> = {};
  for (const advId of Object.keys(rawParams)) {
    out[advId] = { ...rawParams[advId] };
  }
  for (const advId of Object.keys(EXTENT_KEY_RENAMES)) {
    const params = out[advId];
    if (!params) continue;
    const renames = EXTENT_KEY_RENAMES[advId];
    const secondary = SECONDARY[advId] ?? new Set<string>();
    const inverted = INVERTED_PCT_KEYS[advId] ?? new Set<string>();
    // Primaries first, so a secondary alias can never shadow a real value.
    const ordered = Object.keys(renames).sort((a, b) => {
      const sa = secondary.has(a) ? 1 : 0;
      const sb = secondary.has(b) ? 1 : 0;
      return sa - sb || a.localeCompare(b);
    });
    for (const old of ordered) {
      if (!(old in params)) continue;
      const next = renames[old];
      const value = params[old];
      delete params[old];
      if (next in params) continue;  // a newer write under the new key wins
      params[next] = inverted.has(old) ? 100 - value : value;
    }
    if (Object.keys(params).length === 0) delete out[advId];
  }
  return out;
}

File: src/weatherbrief/analysis/advisories/profile_sparsify.py
"""Server-side sparsify of stored profile settings (#405 / #402 slice 2).

The Python sibling of ``web/ts/helpers/profile-sparsify.ts``. Both implement the
same rule so the client's save path and the one-time backfill migration converge
on an identical stored shape: **persist only what differs from the default.**

Absence resolves to the same default at grading time — advisory params via
``evaluate_all``'s ``{**catalog_defaults, **user_params}`` merge, engine methods
via :data:`ENGINE_METHOD_DEFAULTS` — so pruning a value equal to its default is
lossless for grading and lets an already-dense profile shrink.

Three things are swept, and **nothing else is touched**:

1. **Advisory params** (``advisories.params[id][key]``): a stored param is
   dropped only when the catalog declares a default for that exact ``id:key``
   *and* the stored value equals it. A param with no catalog default (unknown /
   legacy key) is kept — we never delete a value we cannot prove is a default.
2. **Legacy ``cloud_method``** (pre-#410 fused ``<style>_<source>`` key): reduced
   to its bare source. NWP (the default) → the key is dropped (absence follows
   the default); DD (**not** the default) → rewritten to the canonical
   ``cloud_source: "dd"`` so the DD grade is preserved exactly. Either way the
   fused key is removed, which is what lets the #410 read-path fallback be
   deleted.
3. **Engine methods** (``icing_method`` / ``cloud_source`` / ``convective_method``):
   dropped when equal to :data:`ENGINE_METHOD_DEFAULTS`; a differing value is a
   deliberate override and is kept.

**``advisories.enabled`` is never touched** (#402): ``fronts`` declares
``default_enabled=False`` yet is injected by default when the
``auto_front_detection`` artifact exists, so an explicit ``fronts: false`` is a
real opt-out that a naive "equals default" rule would wrongly delete. Enable
flags are already centrally switchable via ``resolve_enabled_ids`` and stay out
of the sparsify entirely.
"""

from __future__ import annotations

import copy
from dataclasses import dataclass
from typing import TYPE_CHECKING

from weatherbrief.analysis.advisories.engine_methods import (
    ENGINE_METHOD_DEFAULTS,
    legacy_cloud_source,
)

if TYPE_CHECKING:
    from weatherbrief.models.advisories import AdvisoryCatalogEntry


@dataclass
class SparsifyStats:
    """Per-profile blast radius, summed by the migration for its dry-run report."""

    params_deleted: int = 0
    engine_deleted: int = 0
    cloud_method_dropped: int = 0  # legacy NWP/redundant key removed outright
    cloud_method_rewritten: int = 0  # legacy DD key canonicalized to cloud_source

    @property
    def touched(self) -> bool:
        return bool(
            self.params_deleted
            or self.engine_deleted
            or self.cloud_method_dropped
            or self.cloud_method_rewritten
        )


def build_param_defaults(
    catalog: list[AdvisoryCatalogEntry],
) -> dict[tuple[str, str], float]:
    """Lookup of catalog default keyed ``(advisory_id, param_key)``.

    ``catalog`` is the live ``get_catalog()`` list — imported at migration time
    (never snapshotted as a literal) so a default that moves between merge and
    deploy cannot silently turn the backfill from a no-op into a data change.
    """
    out: dict[tuple[str, str], float] = {}
    for entry in catalog:
        for p in entry.parameters:
            out[(entry.id, p.key)] = p.default
    return out


def sparsify_settings(
    settings: dict,
    param_defaults: dict[tuple[str, str], float],
) -> tuple[dict, SparsifyStats]:
    """Return a sparsified copy of ``settings`` plus a :class:`SparsifyStats`.

    Pure — the input dict is never mutated (the byte-identical test relies on
    comparing the original against the sparsified copy). Order matters: the
    legacy ``cloud_method`` sweep runs *before* the engine-method prune so a
    rewritten ``cloud_source: "dd"`` (an override) survives while a redundant
    ``cloud_source: "nwp"`` (the default) is then dropped.
    """
    out = copy.deepcopy(settings)
    stats = SparsifyStats()

    # 1. Advisory params equal to the catalog default.
    adv = out.get("advisories")
    if isinstance(adv, dict) and isinstance(adv.get("params"), dict):
        pruned: dict[str, dict] = {}
        for adv_id, params in adv["params"].items():
            if not isinstance(params, dict):
                pruned[adv_id] = params
                continue
            kept: dict[str, float] = {}
            for key, val in params.items():
                dflt = param_defaults.get((adv_id, key))
                if dflt is not None and val == dflt:
                    stats.params_deleted += 1
                    continue  # equals default → prune
                kept[key] = val
            if kept:
                pruned[adv_id] = kept
        adv["params"] = pruned

    # 2. Legacy fused ``cloud_method`` → canonical ``cloud_source`` (or dropped).
    if "cloud_method" in out:
        legacy = out.pop("cloud_method")
        if "cloud_source" in out:
            # New key already present and authoritative → the legacy key is
            # redundant regardless of its value.
            stats.cloud_method_dropped += 1
        else:
            source = legacy_cloud_source(legacy)
            if source is not None and source != ENGINE_METHOD_DEFAULTS["cloud_source"]:
                out["cloud_source"] = source  # DD: preserve the override
                stats.cloud_method_rewritten += 1
            else:
                # NWP (the default) or unparseable → absence follows the default.
                stats.cloud_method_dropped += 1

    # 3. Engine methods equal to their declared default.
    for key, dflt in ENGINE_METHOD_DEFAULTS.items():
        if out.get(key) == dflt:
            out.pop(key, None)
            stats.engine_deleted += 1

    return out, stats
